File: examples/franka_trajectory_following/lcm_control_demo.py
from pyexpat.model import XML_CQUANT_NONE
from dairlib import (lcmt_robot_output, lcmt_robot_input, lcmt_c3)

from pydrake.common.yaml import yaml_load
import pydairlib.common
import pydairlib.lcm
from pydairlib.systems import (RobotC3Receiver, 
                               RobotC3Sender,
                               RobotOutputReceiver, RobotOutputSender,
                               LcmOutputDrivenLoop, OutputVector,
                               TimestampedVector,
                               AddActuationRecieverAndStateSenderLcm)

from pydrake.all import (AbstractValue, DiagramBuilder, DrakeLcm, LeafSystem,
                         MultibodyPlant, Parser, RigidTransform, Subscriber,
                         LcmPublisherSystem, TriggerType, AddMultibodyPlantSceneGraph,
                         LcmInterfaceSystem)
#import pydairlib.common

from pydairlib.multibody import (addFlatTerrain, makeNameToPositionsMap, makeNameToVelocitiesMap)
import pydairlib.common
from pydairlib.systems.controllers_franka import C3Controller_franka

import numpy as np

from pydrake.trajectories import PiecewisePolynomial

import math

# load parameters
param = yaml_load(
    filename="examples/franka_trajectory_following/parameters.yaml")
    
lcm = DrakeLcm()

plant = MultibodyPlant(0.0)

#The package addition here seems necessary due to how the URDF is defined
parser = Parser(plant)
parser.package_map().Add("robot_properties_fingers",
                         "examples/trajectory_following/robot_properties_fingers")
parser.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/trajectory_following/robot_properties_fingers/urdf/trifinger_minimal_collision_2.urdf"))
# parser.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
#     "examples/trifinger_simple/robot_properties_fingers/cube/cube_v2.urdf"))
parser.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/trajectory_following/robot_properties_fingers/urdf/sphere.urdf"))


#Fix the base of the finger to the world
X_WI = RigidTransform.Identity()
plant.WeldFrames(plant.world_frame(), plant.GetFrameByName("base_link"), X_WI)
plant.Finalize()

builder = DiagramBuilder()

#############################################################################################


builder_f = DiagramBuilder()
sim_dt = 1e-4
plant_f, scene_graph = AddMultibodyPlantSceneGraph(builder_f, 0.0)
# addFlatTerrain(plant=plant_f, scene_graph=scene_graph, mu_static=1.0,
#                mu_kinetic=1.0)

# The package addition here seems necessary due to how the URDF is defined
parser_f = Parser(plant_f)
parser_f.package_map().Add("robot_properties_fingers",
                         "examples/franka_trajectory_following/robot_properties_fingers")
parser_f.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/franka_trajectory_following/robot_properties_fingers/urdf/trifinger_minimal_collision_2.urdf"))
# parser_f.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
#     "examples/trifinger_simple/robot_properties_fingers/cube/cube_v2.urdf"))
parser_f.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/franka_trajectory_following/robot_properties_fingers/urdf/sphere.urdf"))


# Fix the base of the finger to the world
X_WI_f = RigidTransform.Identity()
plant_f.WeldFrames(plant_f.world_frame(), plant_f.GetFrameByName("base_link"), X_WI_f)
plant_f.Finalize()


#context_f = plant_f.CreateDefaultContext()

plant_ad_f = plant_f.ToAutoDiffXd()

context_ad_f = plant_ad_f.CreateDefaultContext()

diagram_f = builder_f.Build()

diagram_context = diagram_f.CreateDefaultContext()

context_f = diagram_f.GetMutableSubsystemContext(plant_f, diagram_context)

#############################################################################################

builder_franka = DiagramBuilder()
sim_dt = 1e-4
output_dt = 1e-4

plant_franka, scene_graph_franka = AddMultibodyPlantSceneGraph(builder_franka, sim_dt)
# addFlatTerrain(plant=plant, scene_graph=scene_graph, mu_static=1.0,
#               mu_kinetic=1.0)

# The package addition here seems necessary due to how the URDF is defined
parser_franka = Parser(plant_franka)
# parser.package_map().Add("robot_properties_fingers",
#                          "examples/trajectory_following/robot_properties_fingers")
# parser.AddModelFromFile(FindResourceOrThrow(
#     "drake/manipulation/models/franka_description/urdf/panda_arm.urdf"))
parser_franka.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/franka_trajectory_following/robot_properties_fingers/urdf/franka_box.urdf"))
parser_franka.AddModelFromFile(pydairlib.common.FindResourceOrThrow(
    "examples/franka_trajectory_following/robot_properties_fingers/urdf/sphere.urdf"))



#props = mut.ProximityProperties()

#props.AddProperty("material", "point_contact_stiffness", 1000)

# Fix the base of the finger to the world
X_WI_franka = RigidTransform.Identity()
plant_franka.WeldFrames(plant_franka.world_frame(), plant_franka.GetFrameByName("panda_link0"), X_WI_franka)
plant_franka.Finalize()

context_franka = plant_franka.CreateDefaultContext()

state_receiver = builder.AddSystem(RobotOutputReceiver(plant_franka))



nq = plant.num_positions()
nv = plant.num_velocities()
nu = plant.num_actuators()
nc = 2

q = np.zeros((nq,1))
q_map = makeNameToPositionsMap(plant)
v_map = makeNameToVelocitiesMap(plant)

finger_init = param["q_init_finger"]
q[0] = finger_init[0]
q[1] = finger_init[1]
q[2] = finger_init[2]

ball_init = param["q_init_ball_c3"]
q[q_map['base_qw']] = ball_init[0]
q[q_map['base_qx']] = ball_init[1]
q[q_map['base_qy']] = ball_init[2]
q[q_map['base_qz']] = ball_init[3]
q[q_map['base_x']] = ball_init[4]
q[q_map['base_y']] = ball_init[5]
q[q_map['base_z']] = ball_init[6]

mu = param["mu"]

Qinit = param["Q_default"] * np.eye(nq+nv)
Qinit[0,0] = param["Q_finger"]
Qinit[1,1] = param["Q_finger"]
Qinit[2,2] = param["Q_finger"]
Qinit[7,7] = param["Q_ball_x"]
Qinit[8,8] = param["Q_ball_y"]
Qinit[10:10+nv,10:10+nv] = param["Q_ball_vel"] * np.eye(nv)
Qinit[10:13,10:13] = param["Q_finger_vel"]*np.eye(3) #10
# Qinit[13:16, 13:16] = 0*np.eye(3) # zero out cost on rotational velocity
Rinit = param["R"] * np.eye(nu) #torques

#admm_params
Ginit = param["G"] * np.eye(nq+nv+nu+6*nc)
Uinit = param["U_default"] * np.eye(nq+nv+nu+6*nc)
Uinit[0:nq+nv,0:nq+nv] = param["U_pos_vel"] * np.eye(nq+nv) 
Uinit[nq+nv+6*nc:nq+nv+nu+6*nc, nq+nv+6*nc:nq+nv+nu+6*nc] = param["U_u"] * np.eye(nu)

xdesiredinit = np.zeros((nq+nv,1))
xdesiredinit[:nq] = q


r = param["traj_radius"]
x_c = param["x_c"]
y_c = param["y_c"]

degree_increment = param["degree_increment"]
if param["hold_order"] == 0:
    theta = np.arange(degree_increment, 400 + degree_increment, degree_increment)
elif param["hold_order"] == 1:
    theta = np.arange(0, 400, degree_increment)
xtraj = []
for i in theta:
    x = r * np.sin(math.radians(i+param["phase"]))
    y = r * np.cos(math.radians(i+param["phase"]))
    # x = r * np.sin(math.radians(i+param["phase"]+90)) * np.cos(math.radians(i+param["phase"]+90))
    # y = r * np.sin(math.radians(i+param["phase"]+90))
    q[q_map['base_x']] = x + x_c
    q[q_map['base_y']] = y + y_c
    q[q_map['base_z']] = param["ball_radius"]
    xtraj_hold = np.zeros((nq+nv,1))
    xtraj_hold[:nq] = q
    xtraj.append(xtraj_hold)

time_increment = 1.0*param["time_increment"]  # also try just moving in a straight line maybe
delay = param["stabilize_time1"] + param["move_time"] + param["stabilize_time2"]
timings = np.arange(delay, time_increment*len(xtraj) + delay, time_increment)

if param["hold_order"] == 0:
    pp = PiecewisePolynomial.ZeroOrderHold(timings, xtraj)
elif param["hold_order"] == 1:
    pp = PiecewisePolynomial.FirstOrderHold(timings, xtraj)


num_friction_directions = 2
N = 5
Q = []
R = []
G = []
U = []
xdesired = []

for i in range(N):
    Q.append(Qinit)
    R.append(Rinit)
    G.append(Ginit)
    U.append(Uinit)
    xdesired.append(xdesiredinit)

#Qinit[nv-3:nv,nv-3:nv] = 1*np.eye(3) #penalize final velocities
Q.append(Qinit)
xdesired.append(xdesiredinit)

finger_lower_link_0_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("tip_link_1_real"))[0]
#finger_lower_link_120_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("tip_link_2_real"))[0]
#finger_lower_link_240_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("tip_link_3_real"))[0]
#cube_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("cube"))[0]
sphere_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("sphere"))[0]
ground_geoms = plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("box"))[0]

contact_geoms = [finger_lower_link_0_geoms, sphere_geoms, ground_geoms] #finger_lower_link_120_geoms, finger_lower_link_240_geoms,


plant_ad = plant.ToAutoDiffXd()

context = plant.CreateDefaultContext()

context_ad = plant_ad.CreateDefaultContext()

controller = builder.AddSystem(
    C3Controller_franka(plant, plant_f, plant_franka, context, context_f, context_franka, plant_ad, plant_ad_f, context_ad, context_ad_f, scene_graph, diagram_f, contact_geoms, num_friction_directions, mu, Q, R, G, U, xdesired, pp))
state_force_sender = builder.AddSystem(RobotC3Sender(10, 9, 6, 9))


builder.Connect(state_receiver.get_output_port(0), controller.get_input_port(0))

builder.Connect(controller.get_output_port(), state_force_sender.get_input_port(0))

control_publisher = builder.AddSystem(LcmPublisherSystem.Make(
    channel="CONTROLLER_INPUT", lcm_type=lcmt_c3, lcm=lcm,
    publish_triggers={TriggerType.kForced},
    publish_period=0.0, use_cpp_serializer=True))
builder.Connect(state_force_sender.get_output_port(),
    control_publisher.get_input_port())

# TODO: check these connections
# lcm_passthrough = builder.AddSystem(LcmInterfaceSystem(lcm))
# passthrough = AddActuationRecieverAndStateSenderLcm(
#     builder=builder, plant=plant_franka, lcm=lcm_passthrough, actuator_channel="FRANKA_INPUT",
#     state_channel="FRANKA_OUTPUT", publish_rate=1/output_dt,
#     publish_efforts=True, actuator_delay=0.0)   #1/output_dt

diagram = builder.Build()

context_d = diagram.CreateDefaultContext()
receiver_context = diagram.GetMutableSubsystemContext(state_receiver, context_d)

loop = LcmOutputDrivenLoop(drake_lcm=lcm, diagram=diagram,
                          lcm_parser=state_receiver,
                          input_channel="FRANKA_OUTPUT",
                          is_forced_publish=True)

loop.Simulate(200)

File: examples/franka_trajectory_following/lcm_control_demo.cc
#include <vector>
#include <math.h>

#include "drake/systems/analysis/simulator.h"
#include "drake/systems/framework/diagram_builder.h"
#include "drake/systems/lcm/lcm_interface_system.h"
#include "drake/systems/lcm/lcm_publisher_system.h"
#include "drake/systems/lcm/lcm_subscriber_system.h"
#include <drake/lcm/drake_lcm.h>
#include <drake/multibody/tree/multibody_element.h>
#include <drake/multibody/parsing/parser.h>
#include <drake/common/trajectories/piecewise_polynomial.h>
#include <drake/math/rigid_transform.h>
#include "drake/math/autodiff.h"


#include "systems/robot_lcm_systems.h"
#include "dairlib/lcmt_robot_input.hpp"
#include "dairlib/lcmt_robot_output.hpp"
#include "dairlib/lcmt_c3.hpp"
#include "multibody/multibody_utils.h"

#include "c3_parameters.h"
#include "systems/robot_lcm_systems.h"
#include "systems/controllers/c3_controller_franka.h"
#include "systems/framework/lcm_driven_loop.h"

namespace dairlib {
  using drake::geometry::SceneGraph;
  using drake::multibody::MultibodyPlant;
  using drake::multibody::AddMultibodyPlantSceneGraph;
  using drake::math::RigidTransform;
  using drake::systems::DiagramBuilder;
  using drake::systems::lcm::LcmPublisherSystem;
  using drake::systems::lcm::LcmSubscriberSystem;
  using drake::systems::Context;
  using drake::multibody::Parser;
  using multibody::makeNameToPositionsMap;
  using multibody::makeNameToVelocitiesMap;
  using drake::trajectories::PiecewisePolynomial;

  using Eigen::VectorXd;
  using Eigen::MatrixXd;

  int DoMain(int argc, char* argv[]){
    C3Parameters param = drake::yaml::LoadYamlFile<C3Parameters>(
      "examples/franka_trajectory_following/parameters.yaml");
    drake::lcm::DrakeLcm drake_lcm;

    /// parse plant from urdfs
    MultibodyPlant<double> plant(0.0);
    Parser parser(&plant);
    parser.package_map().Add("robot_properties_fingers",
                         "examples/franka_trajectory_following/robot_properties_fingers");
    parser.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/trifinger_minimal_collision_2.urdf");
    parser.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/sphere.urdf");

    /// Fix base of finger to world
    RigidTransform<double> X_WI = RigidTransform<double>::Identity();
    plant.WeldFrames(plant.world_frame(), plant.GetFrameByName("base_link"), X_WI);
    plant.Finalize();

    DiagramBuilder<double> builder;

    /* -------------------------------------------------------------------------------------------*/

    DiagramBuilder<double> builder_f;

    auto [plant_f, scene_graph] = AddMultibodyPlantSceneGraph(&builder_f, 0.0);
    Parser parser_f(&plant_f);
    parser_f.package_map().Add("robot_properties_fingers",
                         "examples/franka_trajectory_following/robot_properties_fingers");
    parser_f.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/trifinger_minimal_collision_2.urdf");
    parser_f.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/sphere.urdf");
    RigidTransform<double> X_WI_f = RigidTransform<double>::Identity();
    plant_f.WeldFrames(plant_f.world_frame(), plant_f.GetFrameByName("base_link"), X_WI_f);
    plant_f.Finalize();

    std::unique_ptr<MultibodyPlant<drake::AutoDiffXd>> plant_ad_f = 
      drake::systems::System<double>::ToAutoDiffXd(plant_f);
    auto context_ad_f = plant_ad_f->CreateDefaultContext();
    auto diagram_f = builder_f.Build();
    std::unique_ptr<Context<double>> diagram_context = diagram_f->CreateDefaultContext();
    auto& context_f = diagram_f->GetMutableSubsystemContext(plant_f, diagram_context.get());

    /* -------------------------------------------------------------------------------------------*/

    DiagramBuilder<double> builder_franka;
    double sim_dt = 1e-4;

    auto [plant_franka, scene_graph_franka] = AddMultibodyPlantSceneGraph(&builder_franka, sim_dt);
    Parser parser_franka(&plant_franka);
    parser_franka.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/franka_box.urdf");
    parser_franka.AddModelFromFile("examples/franka_trajectory_following/robot_properties_fingers/urdf/sphere.urdf");
    RigidTransform<double> X_WI_franka = RigidTransform<double>::Identity();
    plant_franka.WeldFrames(plant_franka.world_frame(), plant_franka.GetFrameByName("panda_link0"), X_WI_franka);
    plant_franka.Finalize();
    auto context_franka = plant_franka.CreateDefaultContext();

    /* -------------------------------------------------------------------------------------------*/

    int nq = plant.num_positions();
    int nv = plant.num_velocities();
    int nu = plant.num_actuators();
    int nc = 2;

    VectorXd q = VectorXd::Zero(nq);
    std::map<std::string, int> q_map = makeNameToPositionsMap(plant);
    std::map<std::string, int> v_map = makeNameToVelocitiesMap(plant);
    
    q(0) = param.q_init_finger(0);
    q(1) = param.q_init_finger(1);
    q(2) = param.q_init_finger(2);

    q[q_map["base_qw"]] = param.q_init_ball_c3(0);
    q[q_map["base_qx"]] = param.q_init_ball_c3(1);
    q[q_map["base_qy"]] = param.q_init_ball_c3(2);
    q[q_map["base_qz"]] = param.q_init_ball_c3(3);
    q[q_map["base_x"]] = param.q_init_ball_c3(4);
    q[q_map["base_y"]] = param.q_init_ball_c3(5);
    q[q_map["base_z"]] = param.q_init_ball_c3(6);
    double mu = param.mu;

    MatrixXd Qinit = param.Q_default * MatrixXd::Identity(nq+nv, nq+nv);
    Qinit.block(0,0,3,3) << param.Q_finger * MatrixXd::Identity(3,3);
    Qinit(7,7) = param.Q_ball_x;
    Qinit(8,8) = param.Q_ball_y;
    Qinit.block(10,10,nv,nv) << param.Q_ball_vel * MatrixXd::Identity(nv,nv);
    Qinit.block(10,10,3,3) << param.Q_finger_vel * MatrixXd::Identity(3,3);
    MatrixXd Rinit = param.R * MatrixXd::Identity(nu, nu);

    MatrixXd Ginit = param.G * MatrixXd::Identity(nq+nv+nu+6*nc, nq+nv+nu+6*nc);
    MatrixXd Uinit = param.U_default * MatrixXd::Identity(nq+nv+nu+6*nc, nq+nv+nu+6*nc);
    Uinit.block(0,0,nq+nv,nq+nv) << 
      param.U_pos_vel * MatrixXd::Identity(nq+nv,nq+nv);
    Uinit.block(nq+nv+6*nc, nq+nv+6*nc, nu, nu) << 
      param.U_u * MatrixXd::Identity(nu, nu);
    
    VectorXd xdesiredinit = VectorXd::Zero(nq+nv);
    xdesiredinit.head(nq) << q;

    /* -------------------------------------------------------------------------------------------*/

    double r = param.traj_radius;
    double xc = param.x_c;
    double yc = param.y_c;
    double degree_increment = param.degree_increment;
    double lower, upper;
    assert(degree_increment > 0); // avoid potential infinite loops

    if (param.hold_order == 0){
      lower = 0;
      upper = 400;
    }
    else if (param.hold_order == 1){
      lower = degree_increment;
      upper = 400 + degree_increment;      
    }
    else{
      std::cout << "Received an invalid hold_order parameter." << std::endl;
      return 0;
    }
    
    std::vector<double> values;
    for (int val = lower; val < upper; val+=degree_increment){
      values.push_back(val);
    }
    VectorXd theta(values.size());
    for (int i = 0; i < (int) values.size(); i++){
      theta(i) = values[i];
    }

    std::vector<VectorXd> xtraj;
    for (int i = 0; i < (int) values.size(); i++){
      double angle = theta(i);
      double x = r * sin(M_PI * (angle + param.phase) / 180.0);
      double y = r * cos(M_PI * (angle + param.phase) / 180.0);
      q(q_map["base_x"]) = x + xc;
      q(q_map["base_y"]) = y + yc;
      q(q_map["base_z"]) = param.ball_radius + param.table_offset;
      VectorXd xtraj_hold = VectorXd::Zero(nq+nv);
      xtraj_hold.head(nq) << q;
      xtraj.push_back(xtraj_hold); 
    }

    // convert xtraj into MatrixXd
    MatrixXd xtraj_mat(values.size(), nq+nv);
    for (int i = 0; i < (int) values.size(); i++){
      xtraj_mat.block(i, 0, 1, nv+nq) << xtraj[i].transpose();
    }

    double time_increment = param.time_increment;
    double settling_time = param.stabilize_time1 + param.move_time + param.stabilize_time2;
    VectorXd timings(values.size());
    for (int i = 0; i < (int) values.size(); i++){
      timings(i) = settling_time + i * time_increment;
    }

    PiecewisePolynomial<double> pp;
    if (param.hold_order == 0){
      pp = PiecewisePolynomial<double>::ZeroOrderHold(timings, xtraj_mat);
    }
    else if (param.hold_order == 1){
      pp = PiecewisePolynomial<double>::FirstOrderHold(timings, xtraj_mat);
    }

    /* -------------------------------------------------------------------------------------------*/

    int num_friction_directions = 2;
    int N = 5;
    std::vector<MatrixXd> Q, R, G, U;
    std::vector<VectorXd> xdesired;

    for (int i = 0; i < N; i++){
      Q.push_back(Qinit);
      R.push_back(Rinit);
      G.push_back(Ginit);
      U.push_back(Uinit);
      xdesired.push_back(xdesiredinit);
    }
    Q.push_back(Qinit);
    xdesired.push_back(xdesiredinit);

    drake::geometry::GeometryId finger_geoms = 
      plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("tip_link_1_real"))[0];
    drake::geometry::GeometryId sphere_geoms = 
      plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("sphere"))[0];
    drake::geometry::GeometryId ground_geoms = 
      plant_f.GetCollisionGeometriesForBody(plant_f.GetBodyByName("box"))[0];
    std::vector<drake::geometry::GeometryId> contact_geoms = 
      {finger_geoms, sphere_geoms, ground_geoms};

    /* -------------------------------------------------------------------------------------------*/

    std::unique_ptr<MultibodyPlant<drake::AutoDiffXd>> plant_ad = 
      drake::systems::System<double>::ToAutoDiffXd(plant);
    auto context = plant.CreateDefaultContext();
    auto context_ad = plant_ad->CreateDefaultContext();

    auto state_receiver = builder.AddSystem<systems::RobotOutputReceiver>(plant_franka);
    
    // TODO: build this line
    // auto controller = builder.AddSystem<systems::controllers::C3Controller_franka>(
    //                                 plant, plant_f, plant_franka, context, 
    //                                 context_f, context_franka, plant_ad, 
    //                                 plant_ad_f, context_ad, context_ad_f, 
    //                                 scene_graph, diagram_f, contact_geoms, 
    //                                 num_friction_directions, mu, Q, R, G, U, 
    //                                 xdesired, pp);
    auto state_force_sender = builder.AddSystem<systems::RobotC3Sender>(10, 9, 6, 9);

    // TODO: build these lines
    // builder.Connect(state_receiver->get_output_port(0), controller->get_input_port(0));    
    // builder.Connect(controller->get_output_port(), state_force_sender->get_input_port(0));

    auto control_publisher = builder.AddSystem(
        LcmPublisherSystem::Make<dairlib::lcmt_c3>(
          "CONTROLLER_INPUT", &drake_lcm, 
          {drake::systems::TriggerType::kForced}, 0.0));

    auto diagram = builder.Build();
    // Run lcm-driven simulation
    systems::LcmDrivenLoop<dairlib::lcmt_robot_output> loop(
        &drake_lcm, std::move(diagram), state_receiver, "FRANKA_OUTPUT", true);
    
    // TODO: code to connect the publisher


    loop.Simulate(std::numeric_limits<double>::infinity());

    return 0;
  }
} // namespace dairlib

int main(int argc, char* argv[]) { dairlib::DoMain(argc, argv);}
